### PR TITLE
Update "How to: Modify string"

### PR DESCRIPTION
**Pull request by BillWagner**
The following updates are part of this PR:

1. Update the samples to use modern C#.
1. Add all sample code to a single downloadable project.
1. Leverage the C# interactive REPL
1. Highlight the immutability of strings in every example.

[internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/how-to/modify-string-contents?branch=pr-en-us-4519)

### CODE REVIEW

File: docs/csharp/how-to/modify-string-contents.md
Comment (pkulikov) [inline docs/csharp/how-to/modify-string-contents.md]: After method list should be *...methods to remove...*

File: docs/csharp/how-to/modify-string-contents.md
Comment (pkulikov) [inline docs/csharp/how-to/modify-string-contents.md]: Like in other How-Tos on strings, let's insert paragraph about where to find the code used in snippets. Then zip-file should be updated?

File: docs/csharp/programming-guide/strings/index.md
Comment (pkulikov) [inline docs/csharp/programming-guide/strings/index.md]: Instead of *Provides a code example...* it better sounds like in a row below on concatenation: *Illustrates...* Not for this PR: description of *How to: Parse Strings...* also might be updated.
Though, description of concatenation is not correct too: string methods are also used for concatenation. I'd review that table as a whole.

File: samples/snippets/csharp/how-to/strings/ModifyStrings.cs
Comment (pkulikov) [inline samples/snippets/csharp/how-to/strings/ModifyStrings.cs]: Shouldn't local methods start with capital letter as well?
Anyway, a good example of local method usage.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: Thanks for the reminder. I wanted to wait until we'd merged the How to search strings PR.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: I did make a few updates here based on your comment.

First, I added "typically" to the first sentence.  I tried introducing the term "unsafe code" there, but I rejected it because the article is not really about using unsafe code, but about how to change the value of strings.

I want to continue to use the term "Modifying strings" because that is a more common search term than "transforming". I agree it is a bit inaccurate, and the content of the article works to make that clear.

Next, I added a bit more explanatory text in the paragraph that introduces the unsafe sample.

I hope that adds the clarity we need.

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: I'd be somewhat more precise here. Fixed doesn't permit access, it just prevents GC from moving the object you might want to access.

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: Links to unsafe, fixed, etc., would also be good here.

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: Nice example!

**Review by rpetrusha — CHANGES_REQUESTED** [review]
I've left a number of comments, @BillWagner. The REPL really makes a difference here, and the pointer example is simple and extremely effective.

File: docs/csharp/how-to/modify-string-contents.md
Comment (pkulikov) [reply docs/csharp/how-to/modify-string-contents.md]: I advocate not to include output as it makes code snippets take more space (my comment as a reader) and harder to maintain (as a contributor). One more thing: when I run and change code, output comment becomes wrong and just takes space between code and actual output (my comment as a user).

File: docs/csharp/how-to/modify-string-contents.md
Comment (pkulikov) [reply docs/csharp/how-to/modify-string-contents.md]: What about the link to how strings are stored (intern)? Or that would be too many implementation details?

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: I re-wrote the opening and moved the discussion of immutable to the last section.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: 😆 This mirrors the split feedback I've gotten when I've shown pages at conferences. I created #4571 to see which people like. Let's see what responses we get.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: I updated everything in this comment, except expanding the sample to include the example method. If I did that, it doesn't run in the REPL.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: Added links for fixed and unsafe. I didn't find a good topic on interned strings though.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: That was already there. Was it originally yours? :clap:

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: This is a much better opening paragraph, Bill! I'd modify the second sentence just a bit to stress that a new string is returned -- "as a new \`string\` object."

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: Rather than "Other methods", "You" to maintain parallelism
No "also" in "You can also treat"
whitespace --> white space (2 words)
YOu has a typo --> You
matches ---> match

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: I'd move this first sentence to after the code example.

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: I'm not sure why everything from the code label on is italicized. Is '_' interpreted as \*\?

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: whitespace --> white space

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: no comma after "local function"

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: No comma after the two occurrences of "Then"

File: docs/csharp/how-to/modify-string-contents.md
Comment (rpetrusha) [inline docs/csharp/how-to/modify-string-contents.md]: It might be useful to link to some information on interning strings: (\[interns\](\<xref:System.String.Intern%2A>))

**Review by rpetrusha — APPROVED** [review]
I've left a number of minor comments, @BillWagner, but this looks really good overall. You can merge once you decide which ones you want to address.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: I think it's the `_`. The italics does not show up on the review site.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: I did a global search for whitespace.

File: docs/csharp/how-to/modify-string-contents.md
Comment (BillWagner) [reply docs/csharp/how-to/modify-string-contents.md]: Good suggestion. I looked and could not find a good topi in our conceptual docs.

**Pull request merged**
